File: python/BJ1654_랜선자르기.py
s=input().split(' ')
kn = [int(i) for i in s]
Ks = [int(input()) for _ in range(kn[0])]
K = kn[0]
N = kn[1]
l = 1
i=0
for i in range(31):
    makable=0
    l=2**i
    for k in Ks:
        makable += k//l
    if makable<N:
        l/=2
        i-=1        
        break
bins = [0]*(i+1)
for j in range(i,-1,-1):
    bins[j]=1
    l=0
    for b in range(i+1):
        if(bins[b]==1):
            l+=2**b
    makable=0
    for k in Ks:
        makable += k//l
    if makable<N:
        bins[j]=0
l=0
for b in range(i+1):
    if(bins[b]==1):
        l+=2**b
print(l)

File: AlgoJava/src/BJ1654_랜선자르기.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class BJ1654_랜선자르기 {

	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		String[] s = br.readLine().split(" ");
		int K = Integer.parseInt(s[0]);
		long N = Long.parseLong(s[1]);
		long[] lines = new long[K];
		for(int k=0;k<K;k++) {
			lines[k] = Long.parseLong(br.readLine());
		}
		long left = 1;
		long right = Integer.MAX_VALUE;
		long mid=Math.round((left+right)/2);
		long makable=0;
		while(left+1<right) {
			mid=Math.round((left+right)/2);
			makable=0;
			for(long k:lines) {
				makable+=k/mid;
			}
			if(makable>=N) {
				left=mid;
			}
			else if(makable<N) {
				right=mid;
			}
		}
		
		bw.write(Long.toString(left)+"\n");
		bw.flush();
		bw.close();
	}

}
